src/h3/_cy/error_system.pyx

```diff
@@ -0,0 +1,235 @@
+error code (aliased as type `H3Error`).
+When these errors happen (and `h3-py` can't recover from them internally),
+they are passed up to the Python/Cython code, where their
+`uint32_t` error values are converted to **custom** Python exception types.
+These custom exception classes all inherit from `H3BaseException`.
+
+There is a 1-1 correspondence between the concrete subclasses of
+`H3BaseException` and the H3 C library `H3ErrorCodes` values.
+The correspondence is intentional, so that the user can refer to the
+H3 C library documentation on these errors.
+
+The (`uint32_t` <-> Exception) correspondence should be clear from
+the names of each error/exception, but the explicit mapping is given by
+a dictionary in the code below.
+
+Note that some "abstract" subclasses of `H3BaseException` are also included to
+group the exceptions by type. (We say "abstract" because Python has no easy
+way to make true abstract exception classes.)
+
+These "abstract" exceptions will never be raised directly by `h3-py`, but they
+allow the user to catch general groups of errors.
+Note that `h3-py` will only ever directly raise
+the "concrete" exception classes.
+
+Summarizing, all exceptions originating from the C library inherit from
+`H3BaseException`, which has both "abstract" and "concrete" subclasses.
+
+**Abstract classes**:
+
+- H3BaseException
+- H3ValueError
+- H3MemoryError
+- H3GridNavigationError
+
+**Concrete classes**:
+
+- H3FailedError
+- H3DomainError
+- H3LatLngDomainError
+- H3ResDomainError
+- H3CellInvalidError
+- H3DirEdgeInvalidError
+- H3UndirEdgeInvalidError
+- H3VertexInvalidError
+- H3PentagonError
+- H3DuplicateInputError
+- H3NotNeighborsError
+- H3ResMismatchError
+- H3MemoryAllocError
+- H3MemoryBoundsError
+- H3OptionInvalidError
+
+
+# TODO: add tests verifying that concrete exception classes have the right error codes associated with them
+"""
+
+from contextlib import contextmanager
+
+from .h3lib cimport (
+    H3Error,
+
+    # H3ErrorCodes enum values
+    E_SUCCESS,
+    E_FAILED,
+    E_DOMAIN,
+    E_LATLNG_DOMAIN,
+    E_RES_DOMAIN,
+    E_CELL_INVALID,
+    E_DIR_EDGE_INVALID,
+    E_UNDIR_EDGE_INVALID,
+    E_VERTEX_INVALID,
+    E_PENTAGON,
+    E_DUPLICATE_INPUT,
+    E_NOT_NEIGHBORS,
+    E_RES_MISMATCH,
+    E_MEMORY,
+    E_MEMORY_BOUNDS,
+    E_OPTION_INVALID,
+)
+
+#
+# Something that should never happen
+#
+class UnknownH3ErrorCode(Exception):
+    """
+    Indicates that the h3-py Python bindings have received an
+    unrecognized error code from the C library.
+
+    This should never happen. Please report if you get this error.
+
+    Note that this exception is *outside* of the
+    H3BaseException class hierarchy.
+    """
+    pass
+
+
+@contextmanager
+def the_error(obj):
+    """
+    Syntactic maple syrup for grouping exception definitions.
+    The associated `with` statement ends up as a not-half-bad
+    approximation to a valid sentence fragment.
```

Comment:
Nit as this does not effect anything about correctness, but this seems a little too cute for my taste. At a minimum I would add an underscore prefix to discourage importing this elsewhere.

Reply:
Definitely guilty of being cute. :)

I'm fine with whatever name suggestions you might have. I mostly just like the code construct for how clear and short
it makes the class hierarchy definitions:

```python
```